### PR TITLE
8282528: AArch64: Incorrect replicate2L_zero rule

### PR DESCRIPTION
**Pull request by theRealELiu**
This patch fixes the wrong matching rule of replicate2L_zero. It was
matched "ReplicateI" by mistake so that long immediates(not only zero)
had to be moved to register first and matched to replicate2L finally. To
fix this trivial bug, this patch fixes the typo and extends the rule of
replicate2L_zero to replicate2L_imm, which now supports all possible
long immediate values.

The final code changes are shown as below:

replicate2L_imm:

        mov   x13, #0xff
        movk  x13, #0xff, lsl #16
        movk  x13, #0xff, lsl #32
        dup   v16.2d, x13

        =>

        movi  v16.2d, #0xff00ff00ff

[Test]
test/jdk/jdk/incubator/vector, test/hotspot/jtreg/compiler/vectorapi
passed without failure.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282528](https://bugs.openjdk.java.net/browse/JDK-8282528): AArch64: Incorrect replicate2L_zero rule

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7939/head:pull/7939` \
`$ git checkout pull/7939`

Update a local copy of the PR: \
`$ git checkout pull/7939` \
`$ git pull https://git.openjdk.java.net/jdk pull/7939/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7939`

View PR using the GUI difftool: \
`$ git pr show -t 7939`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7939.diff">https://git.openjdk.java.net/jdk/pull/7939.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -1332,16 +1332,51 @@ void MacroAssembler::movptr(Register r, uintptr_t imm64) {
...
+    uint64_t one_byte = 0;
+    for (int i = 0; i < 8; i++) {
+      one_byte = tmp & 0xFFULL;
+      if (one_byte != 0xFFULL && one_byte != 0) {
+        can_encode = false;
+        break;
+      }
+      tmp = tmp >> 8;
+    }
+
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1363]: This logic should be in a separate function.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1363]: I think we need a `can_encode(imm, arrangement)` function.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1363]: And then another function that actually does the arranging, and the generation of instructions calls those functions.

Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1363]: Thanks for your review. I agree with that `can_encode(imm, arrangment)` function is better. My concern is that this JBS is just a bug fix for replicate2L_imm backend, and for other SIMD_Arrangment, I found that they can have some other choice for the code generation, but I didn’t touch them in this patch to keep it clear and small.  I list two examples below.

Example1:
```
        movi  v16.4s, #0x34
        orr v16.4s, #0x12, lsl #8

        vs

        mov w8, #0x1234
        dup v16.4s, w8
```

Example2:
```
        movi    v16.4s, #0x78
        orr     v16.4s, #0x56, lsl #8
        orr     v16.4s, #0x34, lsl #16
        orr     v16.4s, #0x12, lsl #24

        vs

        mov     w14, #0x5678
        movk    w14, #0x1234, lsl #16
        dup     v16.4s, w14
```

I'm considering to measure the performance and refine the mov macro assembler if it's necessary. `can_encode` can also be done in the refined work. What do you think?

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1363]: Sure. I'm looking at Neoverse V1 Optimization Guide, which suggests a fairly high cost for core - SIMD moves, and also only a 2 (of 4) of the SIMD pipelines can communicate with the integer registers. So I've got an idea.
Please feel free to do any reorganization later, if you like. It's just that the current organization makes it hard to follow, and thus hard to review.

**Comment by theRealELiu** [comment]
@theRealAph  Could you help to take a look at this?

The latest commit refines the code
generator for macro mov, which now will generate DUP for those
immediates can not be encoded in MOVI. E.g., for the case of
IntVector.broadcast(0x12345678), the final code changes are shown as
below:

Before:
        movi    v16.4s, #0x78
        orr     v16.4s, #0x56, lsl <span>#</span>8
        orr     v16.4s, #0x34, lsl <span>#<span>16
        orr     v16.4s, #0x12, lsl <span>#<span>24

After:
        mov     w14, #0x5678
        movk    w14, #0x1234, lsl  <span>#<span>16
        dup     v16.4s, w14

LLVM also uses DUP for those unencodable immediates which GCC loads from
constant pool.

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4 [hunk 2]
```diff
@@ -1553,55 +1553,50 @@ VFABD(fabd, fabd, 4, F, X, S, 128)
...
+  match(Set dst (Replicate$3 ifelse($4, _imm, con, src)));
+  ins_cost(INSN_COST);
+  format %{ "ifelse($1, mov, movi, $1)  $dst, $ifelse($4, _imm, con, src)`\t# vector ('ifelse($3$4, S_imm, $2H, $2$3)`)"' %}
+  ins_encode %{
+    __ $1(as_FloatRegister($dst$$reg), __ T$2$8,ifelse(
+            `$3 $4', `B _imm', ` '$con$$constant & 0xff,
+            `$3 $4', `S _imm', ` '$con$$constant & 0xffff,
+            `$3 $4', `I _imm', ` '$con$$constant,
+            `$3 $4', `L _imm', ` '$con$$constant,
+            `$6', vRegF,`
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:1571]: This `ifelse` is rather confusing. Can't the '& 0xff' be passed as arguments into `VREPLICATE`?

**Comment by theRealAph** [comment]
It's tricky to reason about this. Sure, a micro-benchmark might suggest that `movz; movk` is better than a bunch of `movi` and `orri` with shifts, but on the other hand `dup(vec, general reg )` takes up slots in the base pipelines that might in non-benchmark code be used for something else. having said that, I don't think the difference is huge, so for now maybe we should just prefer smaller code sizes.

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4 (hunk 2 above)
Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: Done. I think these rules should not have been merged by force. It's hard to maintain.

Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: Thanks for your suggestion. Fixed.

**Review by theRealAph — APPROVED** [review]
Looks good to me.

**Comment by pfustc** [comment]
This has been internally reviewed by me. So I reviewed and sponsored it.
